esp-hal-common/src/interrupt/riscv.rs

```diff
@@ -716,15 +503,307 @@ pub fn _setup_interrupts() {
+        let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+        let intr_map_base = intr.mac_intr_map.as_ptr();
+        intr_map_base
+            .offset(interrupt_number)
+            .write_volatile(cpu_interrupt_number as u32);
+    }
+
+    /// Get interrupt priority
+    #[cfg(not(plic))]
+    #[inline]
+    pub(super) unsafe fn get_priority(cpu_interrupt: CpuInterrupt) -> Priority {
+        let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+        let intr_prio_base = intr.cpu_int_pri_0.as_ptr();
+
+        let prio = intr_prio_base
+            .offset(cpu_interrupt as isize)
+            .read_volatile();
+        core::mem::transmute(prio as u8)
+    }
+
+    /// Get cpu interrupt assigned to peripheral interrupt
+    #[inline]
+    pub(super) unsafe fn get_assigned_cpu_interrupt(interrupt: Interrupt) -> CpuInterrupt {
+        let interrupt_number = interrupt as isize;
+        let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+        let intr_map_base = intr.mac_intr_map.as_ptr();
+
+        let cpu_intr = intr_map_base.offset(interrupt_number).read_volatile();
+
+        core::mem::transmute(cpu_intr)
+    }
+}
+
+#[cfg(plic)]
+mod plic {
+    use super::{CpuInterrupt, InterruptKind, Priority};
+    use crate::{peripherals::Interrupt, Cpu};
+
+    // don't use interrupts reserved for CLIC (0,3,4,7)
+    // for some reason also CPU interrupt 8 doesn't work by default since it's
+    // disabled after reset - so don't use that, too
+    pub(super) const PRIORITY_TO_INTERRUPT: [usize; 15] =
+        [1, 2, 5, 6, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19];
+
+    pub(super) const INTERRUPT_TO_PRIORITY: [usize; 19] = [
+        1, 2, 0, 0, 3, 4, 0, 0, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
+    ];
+
+    const DR_REG_PLIC_MX_BASE: u32 = 0x20001000;
+    const PLIC_MXINT_ENABLE_REG: u32 = DR_REG_PLIC_MX_BASE + 0x0;
+    const PLIC_MXINT_TYPE_REG: u32 = DR_REG_PLIC_MX_BASE + 0x4;
+    const PLIC_MXINT_CLEAR_REG: u32 = DR_REG_PLIC_MX_BASE + 0x8;
+    const PLIC_MXINT0_PRI_REG: u32 = DR_REG_PLIC_MX_BASE + 0x10;
+
+    /// Enable a CPU interrupt
+    pub unsafe fn enable_cpu_interrupt(which: CpuInterrupt) {
+        let cpu_interrupt_number = which as isize;
+        let mxint_enable = PLIC_MXINT_ENABLE_REG as *mut u32;
+        unsafe {
+            mxint_enable.write_volatile(mxint_enable.read_volatile() | 1 << cpu_interrupt_number);
+        }
+    }
+
+    /// Disable the given peripheral interrupt.
+    pub fn disable(_core: Cpu, interrupt: Interrupt) {
+        unsafe {
+            let interrupt_number = interrupt as isize;
+            let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+            let intr_map_base = intr.wifi_mac_intr_map.as_ptr();
+
+            // 0 doesn't disable the pripheral interrupt apparently but will raise a level 1
+            // interrupt - we use CPU interrupt 31 which is disabled after reset by default
+            intr_map_base.offset(interrupt_number).write_volatile(31);
```

Comment:
Are we sure this is the case, seems the TRM disagrees (pg 275), especially this comment:

![image](https://user-images.githubusercontent.com/6977289/224041943-c8d53d74-b2b1-400a-a640-02cd08360c23.png).

Could it be that `wifi_mac_intr_map` is not actually the first interrupt on the esp32c6?

Reply:
It definitely is the first interrupt (see [here](https://github.com/espressif/esp-idf/blob/afbdb0f3ef195ab51690a64e22bfb8a5cd487914/components/soc/esp32c6/interrupts.c#L10)), but I think my original point still stands.

Reply:
Yeah - I've read that sentence a couple of times and tried it even more often
Still can have another look 😄

Reply:
Thing is, the reset value for the mapping registers is 0, so that would mean all interrupt sources are mapped to CPU interrupt 0 by default, which would be quite strange :D. I'll try and look at the esp-idf code too, see how they use it.

Reply:
Ok that is odd - writing 0 to disable _seems_ to work for e.g. TIMG0 but not for WIFI_BB (which is what is tested all the time since that is a problem in e.g. esp-wifi) 🤷‍♂️ that is interesting

Reply:
Ok it seems only that WIFI_BB interrupt is making trouble so I can use 0 here and have the workaround where we need it